### ISSUE TITLE
image_to_tensor with dataloader question

### ISSUE THREAD

**Issue by pablovela5620** [kornia/kornia#276]
I'm a bit confused as to how `kornia.image_to_tensor()` is used with a data loader or if I should just use numpy's transpose method.

As it stands if I load an image in using opencv for my own dataset I get back a numpy array of shape `(HxWxC)`, `kornai.image_to_tensor` will convert this into a tensor of shape `(BxCxHxW)`. This works fine if you're not using pytorch's dataloader, but if you are then you it will automatically convert your numpy array into a torch tensor. 

This mean that this function is not intended to be used inside a data augmentation pipeline? it works so long as I convert to numpy but that seems wasteful when I could just use numpy's transpose function within the dataset

**Comment by edgarriba** [comment]
the behaviour for this function cna be found here: https://github.com/arraiyopensource/kornia/blob/master/kornia/utils/image.py#L29-L37

it depends on the usage you want to make. All the operators in kornia are designed to perform in batch.

**Comment by pablovela5620** [comment]
So let me explain what I'm trying to do. I want to use this for data augmentation and transforms. The issue that I'm facing are as follow. 

I have a custom dataset 
``` python
class DummyDataset(Dataset):
    def __init__(self, data_root=None):
        self.data_root = data_root
        self.image_data = self.data_root[image]
        self.label_data = self.data_root[label]

    def __len__(self):
        return len(self.data_index)

    def __getitem__(self, idx):
        # get data sample
        image_dir = self.image_data[idx]
        label = self.label_data[idx]
        # shape (224, 224, 3) bgr
        image = cv2.imread(image_dir)
        # shape (21,2) 21 keypoints pixel coordinates
        uv = np.array(label_dir)
        return dict(images=image, labels=uv)


# create the dataloader
dataset = DummyDataset()
dataloader = DataLoader(dataset, batch_size=4, shuffle=True)
```
The problem is that first if kornia.image_to_tensor inside the `__getitem__` function then when I use the data loader I'll end up with an extra dimension since image_to_tensor adds in a batch, this could be fixed just using `.squeeze(0)` but then I'll end up with a tensor type and not able to use any torchvision transforms since they are for PIL images. If I try to use it outside the data loader, I am unable since the data loader automatically converts values from np.array into torch.Tensors and image_to_tensor only allows np.ndarray types. The other issue is that I can't use any of pytorchs builtin transforms since it does not support opencv but PIL instead. 

I'm not sure what the pipeline should look like for but I had imaged that I would only do the data loading using opencv inside the __getitem__ and all other transforms in nn.Sequential for both the image data and the label data such as:

```python
image_transform = nn.Sequential(
    # to permute image from (BxHxWxC) to (BxCxHxW)
    kornia.image_to_tensor()
    kornia.BgrToRgb()
    kornia.color.AdjustBrightness(0.5),
    kornia.color.AdjustGamma(gamma=2.),
    kornia.color.AdjustContrast(0.7),
)
```

I'm not sure if the goal is to add things like support for scale/sheer/translation to images as well as for labels (keypoints/bboxes) for data augmentation

**Comment by dvd42** [comment]
> The problem is that first if kornia.image_to_tensor inside the __getitem__ function then when I use the data loader I'll end up with an extra dimension since image_to_tensor adds in a batch, this could be fixed just using .squeeze(0) but then I'll end up with a tensor type and not able to use any torchvision transforms since they are for PIL images

You can try to do `pil_image = Image.fromarray(img.squeeze(0).permute(1, 2, 0).numpy().astype('uint8'))` and then pass these images to the torchvision transform

I am aware that this is a bit of an unclean solution, however...

> I'm not sure if the goal is to add things like support for scale/sheer/translation to images as well as for labels (keypoints/bboxes) for data augmentation

We are actually working on that feature right now. It will be released in the next version and you will no longer need to use PIL Images for data augmentation with torchvion :smiley:

**Comment by pablovela5620** [comment]
Makes sense, thanks! Looking forward to the next version so I can stop using PIL
